Win32-network/test/Test/Async/Socket.hs

```diff
@@ -124,3 +136,112 @@ prop_send_recv (LargeNonEmptyBS bs _size) =
 
           bs' <- takeMVar v
           pure $ bs == bs'
+
+
+socketToBinaryChannel :: Binary a
+                      => Socket
+                      -> BinaryChannel a
+socketToBinaryChannel sock = BinaryChannel { readChannel, writeChannel, closeChannel }
+  where
+    readChannel b = do
+      s <- decode . BL.fromStrict <$> Async.recv sock 8
+      if b
+        then do
+          bs' <- recvLen sock s
+          pure $ Just $ decode bs'
+        else pure Nothing
+
+    writeChannel b a = do
+      let chunks :: [ByteString]
+          chunks = BL.toChunks (encode a)
+          size   :: Int
+          size = bool (+1) id b $ foldl' (\x y -> x + BS.length y) 0 chunks
+      _ <- Async.sendAll sock (BL.toStrict $ encode size)
+      traverse_ (\chunk -> Async.sendAll sock chunk) chunks
+
+    closeChannel = Socket.close sock
+
+
+prop_PingPong :: Positive Int
+              -> Blocking
+              -> LargeNonEmptyBS
+              -> Property
+prop_PingPong (Positive n) blocking (LargeNonEmptyBS bs _bufSize) =
+    ioProperty $ Async.withIOManager $ \iocp ->
+      bracket
+        ((,) <$> Socket.socket Socket.AF_INET Socket.Stream Socket.defaultProtocol
+             <*> Socket.socket Socket.AF_INET Socket.Stream Socket.defaultProtocol)
+        (\(x, y) -> Socket.close x >> Socket.close y)
+        $ \(sockIn, sockOut) -> do
+          lock <- newEmptyMVar
+
+          -- fork a PingPong server
+          let addr = SockAddrInet 3334 (Socket.tupleToHostAddress (127, 0, 0, 1))
```

Comment:
Use an ephemeral port, applies to prop_PingPongPipelined too.

Reply:
Done.  In all of the tests we use now ephemeral ports.